File: clojure/plugins/voxgig/sekreto/plugins/boru.clj
;; A boru vault, as a voxgig/plugin definition.
;;
;; A plugin rather than a built-in kind because it spawns a child process
;; (the CLI) or opens a socket (the wire protocol), and a built-in kind
;; reads at most a local file (docs/design/plugin-providers.md).
;;
;; A port of typescript/plugins/boru.ts, which is canonical.

(ns voxgig.sekreto.plugins.boru
  (:require [clojure.string :as string]
            [voxgig.sekreto.core :as core]
            [voxgig.sekreto.addr :refer [checkaddr]]
            [voxgig.sekreto.json :as json]
            [voxgig.sekreto.provider :as provider]
            [voxgig.sekreto.providers :as providers]
            [voxgig.sekreto.plugins.httpjson :as http]
            [voxgig.sekreto.plugins.proc :as proc]))

(defn borumiss?
  "Does this boru failure mean \"no such secret\" rather than \"I could not
  answer\"? Matched on boru's own wording for a missing alias."
  [why]
  (string/includes? why "no alias named"))

(defrecord Boru [command namespace home addr token mount]
  provider/Provider

  ;; A boru vault (https://github.com/boru-lang/boru).
  ;;
  ;; Two ways in, both boru's own.
  ;;
  ;; With no `addr`, the CLI: `boru vault get --reveal <alias>` prints the
  ;; secret on stdout and nothing else. The passphrase is read by boru
  ;; itself from `BORU_VAULT_PASSPHRASE`; sekreto never accepts it as config
  ;; and never puts it on a command line, where it would show up in the
  ;; process table.
  ;;
  ;; With an `addr`, boru's wire protocol: `boru vault serve` publishes a
  ;; read-only, HashiCorp-shaped provision API (boru's
  ;; design/VAULT-WIRE-PROTOCOL.0.md), authenticated by a capability token
  ;; from `boru vault grant`. A sekreto name is already a valid boru alias,
  ;; and boru aliases keep their dots, so `api.token` is the single path
  ;; segment `api.token` - not the `api`/`token` split a HashiCorp KV gets.
  ;; The value is the `value` field. A 404 is a miss; anything else the
  ;; server refuses (a revoked capability, a sealed vault) is an error.
  ;;
  ;; boru's `vault proxy` and `vault mcp` remain out of bounds: they are a
  ;; credential *broker*, built precisely so the caller never receives the
  ;; credential. `vault serve` is the provision endpoint, built to hand the
  ;; value back - that is the one sekreto uses.
  (lookup [_ name]
    (core/checkname name)

    (if (core/notempty addr)
      (do
        (checkaddr addr)
        ;; The dotted name stays one path segment: boru aliases keep dots.
        (let [alias (if (core/notempty namespace) (str namespace "/" name) name)
              url (str (http/trimslash addr) "/v1/" mount "/data/" alias)
              res (http/fetchjson "GET" url {"X-Vault-Token" (or token "")})]
          (cond
            (= 404 (:status res)) nil

            (not= 200 (:status res))
            (throw (core/sekretoerror (str "sekreto: boru serve error: " (:status res) ": " url)))

            :else (json/astext (json/dig (:body res) "data" "data" "value")))))

      (let [alias (if (core/notempty namespace) (str namespace ":" name) name)
            builder (proc/command [command "vault" "get" "--reveal" alias])]

        (when (core/notempty home)
          (.put (.environment builder) "BORU_HOME" home))

        (let [ran (proc/runcmd builder command)]
          (cond
            ;; boru prints the value and one newline, and nothing else.
            (= 0 (:status ran)) (core/dropsuffix (:out ran) "\n")

            ;; "no alias named" is boru saying it does not hold this secret,
            ;; which is a miss: the chain carries on to the next provider. A
            ;; locked vault or a wrong passphrase is not a miss - treating
            ;; it as one would fall through to a weaker store without saying
            ;; so.
            (borumiss? (:why ran)) nil

            :else
            (throw (core/sekretoerror
                    (str "sekreto: boru vault error: "
                         (if (= "" (:why ran)) (str "exit " (:status ran)) (:why ran))))))))))

  (describe [_]
    (if (core/notempty addr)
      (str "boru:" addr)
      (str "boru" (when (core/notempty namespace) (str ":" namespace))))))

(def boru
  "The `boru` provider kind."
  (providers/providerplugin
   "boru"
   (fn [{:keys [command namespace home addr token mount]}]
     (->Boru (or (core/notempty command) "boru") namespace home
             (some-> (core/notempty addr) http/trimslash) token
             (or (core/notempty mount) "secret")))))

File: java/plugins/com/voxgig/sekreto/plugins/Boru.java
// A boru vault, as a voxgig/plugin definition.
//
// A PLUGIN, NOT PART OF THE CORE: it spawns a process, or opens a
// socket, and the core reads at most a local file. A chain that does not
// name this kind must not carry it, so the calling project imports this
// definition and hands it to Sekreto in the `plugins` option.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.headers;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;
import static com.voxgig.sekreto.plugins.Proc.runcmd;

import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import com.voxgig.sekreto.plugins.Proc.Ran;
import java.util.Map;
import voxgig.plugin.Definition;

/**
 * A boru vault (https://github.com/boru-lang/boru).
 *
 * <p>Two ways in, both boru's own.
 *
 * <p>With no `addr`, the CLI: `boru vault get --reveal <alias>` prints the
 * secret on stdout and nothing else. The passphrase is read by boru itself
 * from `BORU_VAULT_PASSPHRASE`; sekreto never accepts it as config and
 * never puts it on a command line, where it would show up in the process
 * table.
 *
 * <p>With an `addr`, boru's wire protocol: `boru vault serve` publishes a
 * read-only, HashiCorp-shaped provision API (boru's
 * design/VAULT-WIRE-PROTOCOL.0.md), authenticated by a capability token
 * from `boru vault grant`. A sekreto name is already a valid boru alias,
 * and boru aliases keep their dots, so `api.token` is the single path
 * segment `api.token` - not the `api`/`token` split a HashiCorp KV gets.
 * The value is the `value` field. A 404 is a miss; anything else the
 * server refuses (a revoked capability, a sealed vault) is an error.
 *
 * <p>boru's `vault proxy` and `vault mcp` remain out of bounds: they are a
 * credential *broker*, built precisely so the caller never receives the
 * credential. `vault serve` is the provision endpoint, built to hand the
 * value back - that is the one sekreto uses.
 */
public final class Boru implements Provider {

  /** The `boru` kind: what `plugins` hands to Sekreto. */
  public static final Definition PLUGIN = Support.providerplugin("boru", spec ->
      new Boru(
          Support.text(spec.get("command")), Support.text(spec.get("namespace")),
          Support.text(spec.get("home")), Support.text(spec.get("addr")),
          Support.text(spec.get("token")), Support.text(spec.get("mount"))));
  private final String command;
  private final String namespace;
  private final String home;
  private final String addr;
  private final String token;
  private final String mount;

  public Boru(String command, String namespace, String home) {
    this(command, namespace, home, null, null, null);
  }

  public Boru(String command, String namespace, String home,
      String addr, String token, String mount) {
    this.command = null == command || command.isEmpty() ? "boru" : command;
    this.namespace = namespace;
    this.home = home;
    this.addr = null == addr ? "" : trimslash(addr);
    this.token = null == token ? "" : token;
    this.mount = null == mount || mount.isEmpty() ? "secret" : mount;
  }

  @Override
  public String lookup(String name) {
    Sekreto.checkname(name);

    if (!addr.isEmpty()) {
      return wirelookup(name);
    }

    String alias = null == namespace || namespace.isEmpty() ? name : namespace + ":" + name;

    ProcessBuilder builder =
        new ProcessBuilder(command, "vault", "get", "--reveal", alias);

    if (null != home && !home.isEmpty()) {
      builder.environment().put("BORU_HOME", home);
    }

    Ran ran = runcmd(builder, command);
    String out = ran.out;
    String why = ran.why;
    int status = ran.status;

    if (0 == status) {
      // boru prints the value and one newline, and nothing else.
      return out.endsWith("\n") ? out.substring(0, out.length() - 1) : out;
    }

    // "no alias named" is boru saying it does not hold this secret, which is
    // a miss: the chain carries on to the next provider. A locked vault or a
    // wrong passphrase is not a miss - treating it as one would fall through
    // to a weaker store without saying so.
    if (borumiss(why)) {
      return null;
    }

    throw new SekretoError(
        "sekreto: boru vault error: " + (why.isEmpty() ? "exit " + status : why));
  }

  @SuppressWarnings("unchecked")
  private String wirelookup(String name) {
    checkaddr(addr);

    // The dotted name stays one path segment: boru aliases keep their dots.
    String alias = null == namespace || namespace.isEmpty() ? name : namespace + "/" + name;
    String url = addr + "/v1/" + mount + "/data/" + alias;

    Answer res = fetchjson("GET", url, headers("X-Vault-Token", token), null);

    if (404 == res.status) {
      return null;
    }

    if (200 != res.status) {
      throw new SekretoError("sekreto: boru serve error: " + res.status + ": " + url);
    }

    Object data = dig(res.body, "data", "data");
    Object value = data instanceof Map ? ((Map<String, Object>) data).get("value") : null;
    return null == value ? null : String.valueOf(value);
  }

  @Override
  public String describe() {
    if (!addr.isEmpty()) {
      return "boru:" + addr;
    }
    return "boru" + (null == namespace || namespace.isEmpty() ? "" : ":" + namespace);
  }

  /**
   * Does this boru failure mean "no such secret" rather than "I could not
   * answer"? Matched on boru's own wording for a missing alias.
   */
  static boolean borumiss(String why) {
    return why.contains("no alias named");
  }
}
